modules/loggers/mod_log_config.c

```diff
@@ -1666,13 +1782,229 @@ static ap_log_writer *ap_log_set_writer(ap_log_writer *handle)
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                json_empty_string,
+                                r->pool);
+        }
+        else {
+            attribute_value = lfd->portions[i];
+            apr_json_object_set(log_object,
+                                attribute_name, APR_JSON_VALUE_STRING,
+                                apr_json_string_create(r->pool, attribute_value,
+                                                       APR_JSON_VALUE_STRING),
+                                r->pool);
+
+        }
+    }
+
+    ba = apr_bucket_alloc_create(r->pool);
+    bb = apr_brigade_create(r->pool, ba);
+
+    /* encode into bucket brigade */
+    apr_json_encode(bb, NULL, NULL, log_object, APR_JSON_FLAGS_WHITESPACE, r->pool);
+    apr_brigade_puts(bb, NULL, NULL, APR_EOL_STR);
+
+    /* TODO: detect overflow nvec */
+    apr_brigade_to_iovec(bb, vec, &nvec);
+
+    lfdj->portions = apr_palloc(r->pool, sizeof(char*) * nvec);
+    lfdj->lengths = apr_palloc(r->pool, sizeof(int) * nvec);
+    lfdj->total_len = 0;
+    for (i = 0; i < nvec; i++) {
+        lfdj->portions[i] = vec[i].iov_base;
+        lfdj->lengths[i] = vec[i].iov_len;
+        lfdj->total_len += vec[i].iov_len;
+    }
+    lfdj->nelts = nvec;
+    return lfdj;
+}
+#else
+static apr_size_t add_str(apr_array_header_t * ah_strs, apr_array_header_t * ah_strl, const char *str)
+{
+    char** strs = apr_array_push(ah_strs);
+    int* strl = apr_array_push(ah_strl);
+
+    *strs = (char*) str;
+    *strl = strlen(str);
+
+    return *strl;
+}
+
+static ap_log_formatted_data * ap_json_log_formatter( request_rec *r,
+                           void *handle,
+                           ap_log_formatted_data *lfd,
+                           const void *itms)
+
+{
+    json_log_formatter_options * formatter_options = (json_log_formatter_options *) handle;
+    log_format_item *items = (log_format_item *) itms;
+
+    const char* attribute_name;
+    const char* attribute_value;
+
+    int i;
+
+    ap_log_formatted_data *lfdj = apr_palloc(r->pool, sizeof(ap_log_formatted_data));
+    apr_array_header_t *strs = apr_array_make(r->pool, lfd->nelts * 3, sizeof(char *)); /* array of pointers to char */
+    apr_array_header_t *strl = apr_array_make(r->pool, lfd->nelts * 3, sizeof(int));    /* array of int (strlen) */
+
+    lfdj->total_len = 0;
+
+    /* build json object */
+    lfdj->total_len += add_str(strs, strl, "{");
+    for (i = 0; i < lfd->nelts; ++i) {
+        if(items[i].tag == NULL) {
+                continue;
+        }
+
+        if (formatter_options->use_short_attribute_names) {
+                attribute_name = NULL;
+        }
+        else {
+            attribute_name = apr_hash_get(json_hash, items[i].tag, APR_HASH_KEY_STRING);
+        }
+
+        if (!attribute_name) {
+            attribute_name = ap_escape_logjson(r->pool, items[i].tag, NULL, 0); /* use tag as attribute name as fallback */
```

Comment:
The `ap_escape_logjson()` is only needed in the not-apr-1.7 case, as the APR code does all escaping internally.

Reply:
This is the not-apr-util-1.7 case here IIUC.

Reply:
yes, I agree this code is in the apr-util < 1.7 ifdef